File: content/develop/use-cases/feature-store/dotnet/FeatureStore.cs
using System.Collections.Concurrent;
using StackExchange.Redis;

namespace FeatureStoreDemo;

/// <summary>
/// Redis online feature store backed by per-entity Hashes
/// (StackExchange.Redis).
/// </summary>
/// <remarks>
/// Each entity (here, a user) lives at a deterministic key such as
/// <c>fs:user:{id}</c>. The hash holds every feature for that entity
/// as one field per feature — batch-materialized aggregates
/// (refreshed on a daily cycle) alongside streaming-updated signals
/// (refreshed every few seconds). One <c>HMGET</c> returns whichever
/// subset the model needs in one network round trip.
///
/// Two TTL layers solve the <i>mixed staleness</i> problem:
///
/// <list type="bullet">
///   <item>A key-level <c>EXPIRE</c> aligned with the batch
///   materialization cycle.</item>
///   <item>A per-field <c>HEXPIRE</c> on each streaming field gives
///   that field its own shorter expiry, independent of the rest of
///   the hash.</item>
/// </list>
///
/// <c>HEXPIRE</c> and <c>HTTL</c> require Redis 7.4 or later.
/// StackExchange.Redis 2.8+ exposes them as
/// <see cref="IDatabase.HashFieldExpireAsync"/> and
/// <see cref="IDatabase.HashFieldGetTimeToLiveAsync"/>. The demo pins
/// 2.13.17.
///
/// The shared <c>ConnectionMultiplexer</c> is thread-safe and
/// multiplexed — one instance serves the whole process, and every
/// handler in the ASP.NET Core thread pool plus the streaming
/// worker call into it without coordination.
/// </remarks>
public sealed class FeatureStore
{
    public static readonly IReadOnlyList<string> DefaultBatchFields = new[]
    {
        "country_iso",
        "risk_segment",
        "account_age_days",
        "tx_count_7d",
        "avg_amount_30d",
        "chargeback_count_180d",
    };

    public static readonly IReadOnlyList<string> DefaultStreamingFields = new[]
    {
        "last_login_ts",
        "last_device_id",
        "tx_count_5m",
        "failed_logins_15m",
        "session_country",
    };

    public const long DefaultBatchTtlSeconds = 24L * 60L * 60L;
    public const long DefaultStreamingTtlSeconds = 5L * 60L;
    public const string DefaultKeyPrefix = "fs:user:";

    private readonly IConnectionMultiplexer _mux;
    private readonly IDatabase _db;
    public string KeyPrefix { get; }
    public long BatchTtlSeconds { get; }
    public long StreamingTtlSeconds { get; }

    private long _batchWritesTotal;
    private long _streamingWritesTotal;
    private long _readsTotal;
    private long _readFieldsTotal;

    public FeatureStore(
        IConnectionMultiplexer mux,
        string keyPrefix = DefaultKeyPrefix,
        long batchTtlSeconds = DefaultBatchTtlSeconds,
        long streamingTtlSeconds = DefaultStreamingTtlSeconds)
    {
        _mux = mux;
        _db = mux.GetDatabase();
        KeyPrefix = keyPrefix;
        BatchTtlSeconds = batchTtlSeconds;
        StreamingTtlSeconds = streamingTtlSeconds;
    }

    public string KeyFor(string entityId) => KeyPrefix + entityId;

    // ---------------------------------------------------------------
    // Batch ingestion (materialization)
    // ---------------------------------------------------------------

    /// <summary>
    /// Materialize a batch of entities into Redis.
    /// </summary>
    /// <remarks>
    /// One <c>HSET</c> plus one <c>EXPIRE</c> per entity, all queued
    /// through an <c>IBatch</c> so the whole batch ships in a single
    /// network round trip.
    /// </remarks>
    public async Task<int> BulkLoadAsync(
        IReadOnlyDictionary<string, IReadOnlyDictionary<string, object>> rows,
        long ttlSeconds)
    {
        if (rows.Count == 0) return 0;
        var batch = _db.CreateBatch();
        var tasks = new List<Task>(rows.Count * 2);
        foreach (var (entityId, fields) in rows)
        {
            var key = (RedisKey)KeyFor(entityId);
            var entries = new HashEntry[fields.Count];
            int i = 0;
            foreach (var (name, value) in fields)
            {
                entries[i++] = new HashEntry(name, EncodeValue(value));
            }
            tasks.Add(batch.HashSetAsync(key, entries));
            tasks.Add(batch.KeyExpireAsync(key, TimeSpan.FromSeconds(ttlSeconds)));
        }
        batch.Execute();
        await Task.WhenAll(tasks);
        Interlocked.Add(ref _batchWritesTotal, rows.Count);
        return rows.Count;
    }

    // ---------------------------------------------------------------
    // Streaming ingestion
    // ---------------------------------------------------------------

    /// <summary>
    /// Write streaming features with a per-field TTL.
    /// </summary>
    /// <remarks>
    /// <c>HSET</c> and <c>HEXPIRE</c> are queued in the same
    /// <c>IBatch</c> so Redis runs them in pipeline order: the
    /// <c>HSET</c> first creates or overwrites the fields, then
    /// <c>HEXPIRE</c> attaches a TTL to each of those same fields.
    ///
    /// <para>
    /// <see cref="IDatabase.HashFieldExpireAsync"/> returns one
    /// <see cref="ExpireResult"/> per field:
    /// <list type="bullet">
    ///   <item><c>Success</c> (= Redis code 1): TTL set / updated.</item>
    ///   <item><c>Due</c> (= 2): the expiry was 0 or in the past, so
    ///   Redis deleted the field instead of applying a TTL.</item>
    ///   <item><c>ConditionNotMet</c> (= 0): NX/XX/GT/LT condition
    ///   not met (we never use one here).</item>
    ///   <item><c>NoSuchField</c> (= -2): no such field, or no such key.</item>
    /// </list>
    /// We always follow <c>HSET</c> with <c>HEXPIRE</c> so any code
    /// other than <c>Success</c> means the per-field TTL invariant
    /// didn't hold — the helper throws rather than silently leaving a
    /// streaming field with no expiry attached.
    /// </para>
    /// </remarks>
    public async Task UpdateStreamingAsync(
        string entityId,
        IReadOnlyDictionary<string, object> fields,
        long ttlSeconds)
    {
        if (fields.Count == 0) return;
        var key = (RedisKey)KeyFor(entityId);
        var entries = new HashEntry[fields.Count];
        var names = new RedisValue[fields.Count];
        int i = 0;
        foreach (var (name, value) in fields)
        {
            entries[i] = new HashEntry(name, EncodeValue(value));
            names[i] = name;
            i++;
        }

        var batch = _db.CreateBatch();
        var hsetTask = batch.HashSetAsync(key, entries);
        var hexpireTask = batch.HashFieldExpireAsync(
            key, names, TimeSpan.FromSeconds(ttlSeconds));
        batch.Execute();
        await hsetTask;
        var codes = await hexpireTask;
        foreach (var code in codes)
        {
            if (code != ExpireResult.Success)
            {
                throw new InvalidOperationException(
                    $"HEXPIRE did not set every field TTL for {key}: [{string.Join(",", codes)}]");
            }
        }
        Interlocked.Add(ref _streamingWritesTotal, fields.Count);
    }

    // ---------------------------------------------------------------
    // Inference reads
    // ---------------------------------------------------------------

    /// <summary>
    /// Retrieve a subset of features for one entity with <c>HMGET</c>.
    /// Returns only the fields that actually exist on the hash;
    /// missing fields are dropped from the result.
    /// </summary>
    public async Task<Dictionary<string, string>> GetFeaturesAsync(
        string entityId, IReadOnlyList<string> fieldNames)
    {
        var key = (RedisKey)KeyFor(entityId);
        var out_ = new Dictionary<string, string>();
        if (fieldNames.Count == 0) return out_;
        var values = await _db.HashGetAsync(
            key, fieldNames.Select(f => (RedisValue)f).ToArray());
        for (int i = 0; i < fieldNames.Count; i++)
        {
            if (!values[i].IsNull)
            {
                out_[fieldNames[i]] = values[i].ToString();
            }
        }
        Interlocked.Increment(ref _readsTotal);
        Interlocked.Add(ref _readFieldsTotal, out_.Count);
        return out_;
    }

    /// <summary>
    /// Full-hash read via <c>HGETALL</c>. Useful for debugging but
    /// the model server should always go through
    /// <see cref="GetFeaturesAsync"/> with an explicit field list.
    /// </summary>
    public async Task<Dictionary<string, string>> GetAllFeaturesAsync(string entityId)
    {
        var entries = await _db.HashGetAllAsync(KeyFor(entityId));
        var dict = new Dictionary<string, string>(entries.Length);
        foreach (var e in entries)
        {
            dict[e.Name.ToString()] = e.Value.ToString();
        }
        Interlocked.Increment(ref _readsTotal);
        Interlocked.Add(ref _readFieldsTotal, entries.Length);
        return dict;
    }

    /// <summary>
    /// Pipeline <c>HMGET</c> across many entities for batch scoring.
    /// One round trip for the whole batch via <c>IBatch</c>.
    /// </summary>
    public async Task<Dictionary<string, Dictionary<string, string>>> BatchGetFeaturesAsync(
        IReadOnlyList<string> entityIds, IReadOnlyList<string> fieldNames)
    {
        if (entityIds.Count == 0 || fieldNames.Count == 0)
            return new Dictionary<string, Dictionary<string, string>>();

        var fieldValues = fieldNames.Select(f => (RedisValue)f).ToArray();
        var batch = _db.CreateBatch();
        var tasks = new Task<RedisValue[]>[entityIds.Count];
        for (int i = 0; i < entityIds.Count; i++)
        {
            tasks[i] = batch.HashGetAsync(KeyFor(entityIds[i]), fieldValues);
        }
        batch.Execute();
        var rows = await Task.WhenAll(tasks);

        var out_ = new Dictionary<string, Dictionary<string, string>>();
        long seen = 0;
        for (int i = 0; i < entityIds.Count; i++)
        {
            var row = new Dictionary<string, string>();
            for (int j = 0; j < fieldNames.Count; j++)
            {
                if (!rows[i][j].IsNull)
                {
                    row[fieldNames[j]] = rows[i][j].ToString();
                    seen++;
                }
            }
            out_[entityIds[i]] = row;
        }
        Interlocked.Add(ref _readsTotal, entityIds.Count);
        Interlocked.Add(ref _readFieldsTotal, seen);
        return out_;
    }

    // ---------------------------------------------------------------
    // TTL inspection (used by the demo UI)
    // ---------------------------------------------------------------

    /// <summary>
    /// Seconds until the entity key expires. Returns -1 if no TTL is
    /// set, -2 if the key doesn't exist.
    /// </summary>
    public async Task<long> KeyTtlSecondsAsync(string entityId)
    {
        var ttl = await _db.KeyTimeToLiveAsync(KeyFor(entityId));
        if (ttl == null)
        {
            // StackExchange.Redis returns null both for "no TTL" and
            // for "key doesn't exist". Disambiguate with KeyExists.
            return await _db.KeyExistsAsync(KeyFor(entityId)) ? -1L : -2L;
        }
        return (long)ttl.Value.TotalSeconds;
    }

    /// <summary>
    /// Per-field TTL helper (Redis 7.4+). Returns whole seconds for
    /// parity with the other clients: positive seconds remaining,
    /// -1 no field TTL, -2 field (or key) missing.
    /// </summary>
    /// <remarks>
    /// <c>HashFieldGetTimeToLiveAsync</c> wraps <c>HPTTL</c> (not
    /// <c>HTTL</c>), so the API returns milliseconds. We convert to
    /// whole seconds here so the JSON shape matches Python, Node.js,
    /// Go, Java, Rust, Ruby, and PHP, which all expose seconds.
    /// </remarks>
    public async Task<Dictionary<string, long>> FieldTtlsSecondsAsync(
        string entityId, IReadOnlyList<string> fieldNames)
    {
        var out_ = new Dictionary<string, long>();
        if (fieldNames.Count == 0) return out_;
        var values = fieldNames.Select(f => (RedisValue)f).ToArray();
        var ms = await _db.HashFieldGetTimeToLiveAsync(KeyFor(entityId), values);
        // SE.Redis 2.13 returns a flat long[] of length == fieldNames.Count
        // (filled with -2s for a missing key). Coerce defensively against
        // any future version that might return a shorter or empty array.
        for (int i = 0; i < fieldNames.Count; i++)
        {
            // HPTTL returns ms remaining; negative sentinels pass
            // through. Convert positive durations to whole seconds
            // for parity with the other clients' helpers.
            long v = i < ms.Length ? ms[i] : -2L;
            out_[fieldNames[i]] = v < 0 ? v : v / 1000;
        }
        return out_;
    }

    // ---------------------------------------------------------------
    // Demo housekeeping
    // ---------------------------------------------------------------

    /// <summary>
    /// Enumerate up to <paramref name="limit"/> entity IDs by
    /// scanning <c>keyPrefix*</c>. <c>SCAN</c> is non-blocking; the
    /// demo uses it for UI dropdowns, not as a serving primitive.
    /// </summary>
    public List<string> ListEntityIds(int limit)
    {
        var ids = new List<string>(Math.Min(limit, 1024));
        foreach (var endPoint in _mux.GetEndPoints())
        {
            var server = _mux.GetServer(endPoint);
            // pageSize=200 mirrors the other clients' SCAN COUNT
            foreach (var key in server.Keys(
                pattern: KeyPrefix + "*", pageSize: 200))
            {
                var k = key.ToString();
                if (k.Length > KeyPrefix.Length)
                {
                    ids.Add(k[KeyPrefix.Length..]);
                    if (ids.Count >= limit) break;
                }
            }
            if (ids.Count >= limit) break;
        }
        ids.Sort(StringComparer.Ordinal);
        return ids;
    }

    /// <summary>
    /// Count every entity under the key prefix. Iterates SCAN without
    /// an in-memory cap so the UI can show the true total even when
    /// more keys exist than <see cref="ListEntityIds"/> returns.
    /// </summary>
    public long CountEntities()
    {
        long count = 0;
        foreach (var endPoint in _mux.GetEndPoints())
        {
            var server = _mux.GetServer(endPoint);
            foreach (var _ in server.Keys(
                pattern: KeyPrefix + "*", pageSize: 500))
            {
                count++;
            }
        }
        return count;
    }

    public Task<long> DeleteEntityAsync(string entityId) =>
        _db.KeyDeleteAsync(KeyFor(entityId)).ContinueWith(t => t.Result ? 1L : 0L);

    /// <summary>
    /// Drop every entity under the key prefix. Used by the demo
    /// reset path; SCANs and DELs in batches of 500.
    /// </summary>
    public async Task<long> ResetAsync()
    {
        long deleted = 0;
        foreach (var endPoint in _mux.GetEndPoints())
        {
            var server = _mux.GetServer(endPoint);
            var batch = new List<RedisKey>(500);
            foreach (var key in server.Keys(
                pattern: KeyPrefix + "*", pageSize: 500))
            {
                batch.Add(key);
                if (batch.Count >= 500)
                {
                    deleted += await _db.KeyDeleteAsync(batch.ToArray());
                    batch.Clear();
                }
            }
            if (batch.Count > 0)
            {
                deleted += await _db.KeyDeleteAsync(batch.ToArray());
            }
        }
        return deleted;
    }

    public Stats StatsSnapshot() => new(
        Interlocked.Read(ref _batchWritesTotal),
        Interlocked.Read(ref _streamingWritesTotal),
        Interlocked.Read(ref _readsTotal),
        Interlocked.Read(ref _readFieldsTotal));

    public void ResetStats()
    {
        Interlocked.Exchange(ref _batchWritesTotal, 0);
        Interlocked.Exchange(ref _streamingWritesTotal, 0);
        Interlocked.Exchange(ref _readsTotal, 0);
        Interlocked.Exchange(ref _readFieldsTotal, 0);
    }

    public record Stats(
        long BatchWritesTotal,
        long StreamingWritesTotal,
        long ReadsTotal,
        long ReadFieldsTotal);

    /// <summary>
    /// Render a feature value as a string for hash storage. Booleans
    /// become "true"/"false" so they round-trip cleanly through other
    /// clients and redis-cli.
    /// </summary>
    public static string EncodeValue(object? value) => value switch
    {
        null => "",
        bool b => b ? "true" : "false",
        double d when d == Math.Floor(d) => d.ToString("F1", System.Globalization.CultureInfo.InvariantCulture),
        double d => d.ToString(System.Globalization.CultureInfo.InvariantCulture),
        float f => f.ToString(System.Globalization.CultureInfo.InvariantCulture),
        _ => value.ToString() ?? "",
    };
}

File: content/develop/use-cases/feature-store/rust/feature_store.rs
//! `FeatureStore` async helper around a `redis::aio::ConnectionManager`.
//!
//! ConnectionManager is the canonical "single-connection across many
//! async tasks" abstraction in redis-rs: it owns the multiplexed
//! `MultiplexedConnection` underneath and transparently reconnects on
//! a closed socket. The struct holds `Clone` handles, so wiring one
//! into HTTP handlers and a streaming worker is just `clone()`s.

use std::collections::BTreeMap;
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Arc;

use redis::aio::ConnectionManager;
use redis::{AsyncCommands, FromRedisValue, RedisResult, Value};
use serde::Serialize;

/// Default batch feature schema. Daily aggregates computed offline
/// and bulk-loaded once per materialization cycle.
pub const DEFAULT_BATCH_FIELDS: &[&str] = &[
    "country_iso",
    "risk_segment",
    "account_age_days",
    "tx_count_7d",
    "avg_amount_30d",
    "chargeback_count_180d",
];

/// Default streaming feature schema. Updated by the streaming worker
/// as new events arrive, with a per-field TTL so each field
/// self-expires when its upstream pipeline stops.
pub const DEFAULT_STREAMING_FIELDS: &[&str] = &[
    "last_login_ts",
    "last_device_id",
    "tx_count_5m",
    "failed_logins_15m",
    "session_country",
];

pub const DEFAULT_BATCH_TTL_SECONDS: u64 = 24 * 60 * 60;
pub const DEFAULT_STREAMING_TTL_SECONDS: u64 = 5 * 60;
pub const DEFAULT_KEY_PREFIX: &str = "fs:user:";

/// One feature value the helper knows how to encode.
#[derive(Debug, Clone)]
pub enum FeatureValue {
    Str(String),
    Bool(bool),
    Int(i64),
    Float(f64),
}

impl FeatureValue {
    /// Render the value as the string Redis hash fields require.
    /// Booleans become "true"/"false" so they round-trip cleanly
    /// through redis-cli and the other clients.
    pub fn encode(&self) -> String {
        match self {
            FeatureValue::Str(s) => s.clone(),
            FeatureValue::Bool(b) => (if *b { "true" } else { "false" }).into(),
            FeatureValue::Int(i) => i.to_string(),
            FeatureValue::Float(f) => {
                if f.fract() == 0.0 {
                    format!("{:.1}", f)
                } else {
                    f.to_string()
                }
            }
        }
    }
}

impl From<&str> for FeatureValue {
    fn from(s: &str) -> Self { FeatureValue::Str(s.into()) }
}
impl From<String> for FeatureValue {
    fn from(s: String) -> Self { FeatureValue::Str(s) }
}
impl From<bool> for FeatureValue {
    fn from(b: bool) -> Self { FeatureValue::Bool(b) }
}
impl From<i32> for FeatureValue {
    fn from(i: i32) -> Self { FeatureValue::Int(i as i64) }
}
impl From<i64> for FeatureValue {
    fn from(i: i64) -> Self { FeatureValue::Int(i) }
}
impl From<u64> for FeatureValue {
    fn from(i: u64) -> Self { FeatureValue::Int(i as i64) }
}
impl From<f64> for FeatureValue {
    fn from(f: f64) -> Self { FeatureValue::Float(f) }
}

/// One row of features for one entity. Insertion order is preserved
/// so the demo's debug views read in a stable order.
pub type FeatureMap = BTreeMap<String, FeatureValue>;

#[derive(Debug, Clone, Serialize)]
pub struct Stats {
    pub batch_writes_total: i64,
    pub streaming_writes_total: i64,
    pub reads_total: i64,
    pub read_fields_total: i64,
}

#[derive(Clone)]
pub struct FeatureStore {
    conn: ConnectionManager,
    key_prefix: String,
    batch_ttl_seconds: u64,
    streaming_ttl_seconds: u64,
    counters: Arc<Counters>,
}

struct Counters {
    batch_writes_total: AtomicI64,
    streaming_writes_total: AtomicI64,
    reads_total: AtomicI64,
    read_fields_total: AtomicI64,
}

impl FeatureStore {
    pub fn new(
        conn: ConnectionManager,
        key_prefix: impl Into<String>,
        batch_ttl_seconds: u64,
        streaming_ttl_seconds: u64,
    ) -> Self {
        Self {
            conn,
            key_prefix: key_prefix.into(),
            batch_ttl_seconds,
            streaming_ttl_seconds,
            counters: Arc::new(Counters {
                batch_writes_total: AtomicI64::new(0),
                streaming_writes_total: AtomicI64::new(0),
                reads_total: AtomicI64::new(0),
                read_fields_total: AtomicI64::new(0),
            }),
        }
    }

    pub fn key_prefix(&self) -> &str { &self.key_prefix }
    pub fn batch_ttl_seconds(&self) -> u64 { self.batch_ttl_seconds }
    pub fn streaming_ttl_seconds(&self) -> u64 { self.streaming_ttl_seconds }

    pub fn key_for(&self, entity_id: &str) -> String {
        format!("{}{}", self.key_prefix, entity_id)
    }

    // ---------------------------------------------------------------
    // Batch ingestion (materialization)
    // ---------------------------------------------------------------

    /// Materialize a batch of entities into Redis.
    ///
    /// One `HSET` plus one `EXPIRE` per entity, all queued through a
    /// non-transactional pipeline so the whole batch ships in a
    /// single network round trip.
    pub async fn bulk_load(
        &self,
        rows: &[(String, FeatureMap)],
        ttl_seconds: u64,
    ) -> RedisResult<usize> {
        if rows.is_empty() { return Ok(0); }
        let mut pipe = redis::pipe();
        for (entity_id, fields) in rows {
            let key = self.key_for(entity_id);
            let encoded: Vec<(&str, String)> = fields
                .iter()
                .map(|(k, v)| (k.as_str(), v.encode()))
                .collect();
            pipe.hset_multiple(&key, &encoded).ignore();
            pipe.expire(&key, ttl_seconds as i64).ignore();
        }
        let mut conn = self.conn.clone();
        pipe.query_async::<()>(&mut conn).await?;
        self.counters
            .batch_writes_total
            .fetch_add(rows.len() as i64, Ordering::Relaxed);
        Ok(rows.len())
    }

    pub async fn update_batch_feature(
        &self,
        entity_id: &str,
        field: &str,
        value: FeatureValue,
    ) -> RedisResult<()> {
        let mut conn = self.conn.clone();
        let _: () = conn.hset(self.key_for(entity_id), field, value.encode()).await?;
        self.counters
            .batch_writes_total
            .fetch_add(1, Ordering::Relaxed);
        Ok(())
    }

    // ---------------------------------------------------------------
    // Streaming ingestion
    // ---------------------------------------------------------------

    /// Write streaming features with a per-field TTL.
    ///
    /// `HSET` + `HEXPIRE` are queued in a single pipeline so Redis
    /// runs them in order on the server: the `HSET` first creates or
    /// overwrites the fields, then `HEXPIRE` attaches a TTL to each
    /// of those same fields.
    ///
    /// `HEXPIRE` returns one status code per field:
    /// `1` = TTL set, `2` = the expiry was 0 or in the past (so Redis
    /// deleted the field instead), `0` = an NX/XX/GT/LT conditional
    /// flag wasn't met, `-2` = no such field or key. We always
    /// follow `HSET` with `HEXPIRE` so any code other than `1` means
    /// the per-field TTL invariant didn't hold — the helper returns
    /// a `RedisError` rather than silently leaving a streaming field
    /// without an expiry attached.
    pub async fn update_streaming(
        &self,
        entity_id: &str,
        fields: &FeatureMap,
        ttl_seconds: u64,
    ) -> RedisResult<()> {
        if fields.is_empty() { return Ok(()); }
        let key = self.key_for(entity_id);
        let encoded: Vec<(&str, String)> = fields
            .iter()
            .map(|(k, v)| (k.as_str(), v.encode()))
            .collect();
        let names: Vec<&str> = fields.keys().map(|s| s.as_str()).collect();

        let mut pipe = redis::pipe();
        pipe.hset_multiple(&key, &encoded).ignore();
        // HEXPIRE wire form: HEXPIRE key seconds FIELDS count field...
        let mut hexpire = redis::cmd("HEXPIRE");
        hexpire.arg(&key).arg(ttl_seconds).arg("FIELDS").arg(names.len());
        for n in &names { hexpire.arg(n); }
        pipe.add_command(hexpire);

        let mut conn = self.conn.clone();
        // The pipeline returns one entry per non-ignored command;
        // HSET's reply was dropped with .ignore() above, so there is
        // exactly one entry left — HEXPIRE's per-field code list
        // (itself a Vec<i64>). We unwrap the outer Vec to get at
        // those codes.
        let pipe_result: Vec<Vec<i64>> = pipe.query_async(&mut conn).await?;
        let codes = pipe_result.into_iter().next().unwrap_or_default();
        for code in &codes {
            if *code != 1 {
                return Err(redis::RedisError::from((
                    redis::ErrorKind::ResponseError,
                    "HEXPIRE invariant violated",
                    format!(
                        "HEXPIRE did not set every field TTL for {key}: {codes:?}"
                    ),
                )));
            }
        }
        self.counters
            .streaming_writes_total
            .fetch_add(fields.len() as i64, Ordering::Relaxed);
        Ok(())
    }

    // ---------------------------------------------------------------
    // Inference reads
    // ---------------------------------------------------------------

    /// Retrieve a subset of features for one entity with `HMGET`.
    /// Returns only the fields that actually exist on the hash —
    /// missing fields are dropped from the result.
    pub async fn get_features(
        &self,
        entity_id: &str,
        field_names: &[&str],
    ) -> RedisResult<BTreeMap<String, String>> {
        if field_names.is_empty() {
            return Ok(BTreeMap::new());
        }
        let key = self.key_for(entity_id);
        let mut conn = self.conn.clone();
        let values: Vec<Option<String>> = conn.hget(&key, field_names).await?;
        let mut out = BTreeMap::new();
        for (n, v) in field_names.iter().zip(values.into_iter()) {
            if let Some(s) = v {
                out.insert((*n).to_string(), s);
            }
        }
        self.counters.reads_total.fetch_add(1, Ordering::Relaxed);
        self.counters
            .read_fields_total
            .fetch_add(out.len() as i64, Ordering::Relaxed);
        Ok(out)
    }

    /// Full-hash read via `HGETALL`. Useful for debugging but the
    /// model server should always go through `get_features` with an
    /// explicit field list.
    pub async fn get_all_features(
        &self,
        entity_id: &str,
    ) -> RedisResult<BTreeMap<String, String>> {
        let mut conn = self.conn.clone();
        let map: BTreeMap<String, String> = conn.hgetall(self.key_for(entity_id)).await?;
        self.counters.reads_total.fetch_add(1, Ordering::Relaxed);
        self.counters
            .read_fields_total
            .fetch_add(map.len() as i64, Ordering::Relaxed);
        Ok(map)
    }

    /// Pipeline `HMGET` across many entities for batch scoring. One
    /// round trip for the whole batch.
    pub async fn batch_get_features(
        &self,
        entity_ids: &[String],
        field_names: &[&str],
    ) -> RedisResult<BTreeMap<String, BTreeMap<String, String>>> {
        if entity_ids.is_empty() || field_names.is_empty() {
            return Ok(BTreeMap::new());
        }
        let mut pipe = redis::pipe();
        for id in entity_ids {
            pipe.hget(self.key_for(id), field_names);
        }
        let mut conn = self.conn.clone();
        let rows: Vec<Vec<Option<String>>> = pipe.query_async(&mut conn).await?;

        let mut out = BTreeMap::new();
        let mut seen = 0i64;
        for (id, values) in entity_ids.iter().zip(rows.into_iter()) {
            let mut row = BTreeMap::new();
            for (n, v) in field_names.iter().zip(values.into_iter()) {
                if let Some(s) = v {
                    row.insert((*n).to_string(), s);
                    seen += 1;
                }
            }
            out.insert(id.clone(), row);
        }
        self.counters
            .reads_total
            .fetch_add(entity_ids.len() as i64, Ordering::Relaxed);
        self.counters.read_fields_total.fetch_add(seen, Ordering::Relaxed);
        Ok(out)
    }

    // ---------------------------------------------------------------
    // TTL inspection (used by the demo UI)
    // ---------------------------------------------------------------

    /// Seconds until the entity key expires: positive means TTL
    /// remaining, `-1` means no key-level TTL set, `-2` means the
    /// key doesn't exist.
    pub async fn key_ttl_seconds(&self, entity_id: &str) -> RedisResult<i64> {
        let mut conn = self.conn.clone();
        conn.ttl(self.key_for(entity_id)).await
    }

    /// Per-field TTL via `HTTL` (Redis 7.4+). Each value mirrors
    /// `TTL`'s convention: positive seconds remaining, `-1` no field
    /// TTL, `-2` field (or key) missing.
    pub async fn field_ttls_seconds(
        &self,
        entity_id: &str,
        field_names: &[&str],
    ) -> RedisResult<BTreeMap<String, i64>> {
        if field_names.is_empty() {
            return Ok(BTreeMap::new());
        }
        let key = self.key_for(entity_id);
        let mut cmd = redis::cmd("HTTL");
        cmd.arg(&key).arg("FIELDS").arg(field_names.len());
        for n in field_names { cmd.arg(*n); }
        let mut conn = self.conn.clone();
        // HTTL on a missing key still returns a flat list of -2s
        // (one per requested field), so we don't need a defensive
        // shape coercion here the way redis-py and Lettuce do for
        // their respective clients.
        let values: Value = cmd.query_async(&mut conn).await?;
        let codes: Vec<i64> = Vec::<i64>::from_redis_value(&values).unwrap_or_else(|_| {
            // Defensive fallback: if the client ever returns nil for
            // a missing key, treat every field as -2.
            field_names.iter().map(|_| -2).collect()
        });
        let mut out = BTreeMap::new();
        for (n, v) in field_names.iter().zip(codes.into_iter()) {
            out.insert((*n).to_string(), v);
        }
        Ok(out)
    }

    // ---------------------------------------------------------------
    // Demo housekeeping
    // ---------------------------------------------------------------

    /// Enumerate up to `limit` entity IDs by scanning `keyPrefix*`.
    /// `SCAN` is non-blocking; the demo uses it for UI dropdowns,
    /// not as a serving primitive. Result is sorted.
    pub async fn list_entity_ids(&self, limit: usize) -> RedisResult<Vec<String>> {
        use redis::AsyncIter;
        let pattern = format!("{}*", self.key_prefix);
        let mut conn = self.conn.clone();
        let mut iter: AsyncIter<String> = conn.scan_match(&pattern).await?;
        let mut ids: Vec<String> = Vec::new();
        let prefix_len = self.key_prefix.len();
        while let Some(k) = iter.next_item().await {
            if k.len() > prefix_len {
                ids.push(k[prefix_len..].to_string());
            }
            if ids.len() >= limit { break; }
        }
        ids.sort();
        Ok(ids)
    }

    /// Count every entity under the key prefix. Loops `SCAN` without
    /// an in-memory cap so the UI can show the true total even when
    /// more keys exist than `list_entity_ids` returns.
    pub async fn count_entities(&self) -> RedisResult<i64> {
        use redis::AsyncIter;
        let pattern = format!("{}*", self.key_prefix);
        let mut conn = self.conn.clone();
        let mut iter: AsyncIter<String> = conn.scan_match(&pattern).await?;
        let mut n = 0i64;
        while let Some(_k) = iter.next_item().await {
            n += 1;
        }
        Ok(n)
    }

    pub async fn delete_entity(&self, entity_id: &str) -> RedisResult<i64> {
        let mut conn = self.conn.clone();
        conn.del(self.key_for(entity_id)).await
    }

    /// Drop every entity under the key prefix. Used by the demo
    /// reset path; collects keys with `SCAN` then issues variadic
    /// `DEL` in batches of 500.
    pub async fn reset(&self) -> RedisResult<i64> {
        use redis::AsyncIter;
        let pattern = format!("{}*", self.key_prefix);
        let mut conn = self.conn.clone();
        let mut deleted = 0i64;
        let mut batch: Vec<String> = Vec::with_capacity(500);
        let mut iter: AsyncIter<String> = conn.scan_match(&pattern).await?;
        while let Some(k) = iter.next_item().await {
            batch.push(k);
            if batch.len() >= 500 {
                let n: i64 = self.conn.clone().del(batch.as_slice()).await?;
                deleted += n;
                batch.clear();
            }
        }
        if !batch.is_empty() {
            let n: i64 = self.conn.clone().del(batch.as_slice()).await?;
            deleted += n;
        }
        Ok(deleted)
    }

    pub fn stats(&self) -> Stats {
        Stats {
            batch_writes_total: self.counters.batch_writes_total.load(Ordering::Relaxed),
            streaming_writes_total: self.counters.streaming_writes_total.load(Ordering::Relaxed),
            reads_total: self.counters.reads_total.load(Ordering::Relaxed),
            read_fields_total: self.counters.read_fields_total.load(Ordering::Relaxed),
        }
    }

    pub fn reset_stats(&self) {
        self.counters.batch_writes_total.store(0, Ordering::Relaxed);
        self.counters.streaming_writes_total.store(0, Ordering::Relaxed);
        self.counters.reads_total.store(0, Ordering::Relaxed);
        self.counters.read_fields_total.store(0, Ordering::Relaxed);
    }
}
